### ISSUE TITLE
Backprop-able indexing

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#487]
Indexing is a fundamental feature of array libraries. Why not make Variable able to index? Following features are already supported by both NumPy and CuPy:
- Basic indexing (element, slice, newaxis, ellipsis)
- take function (fancy indexing on one axis)

**Comment by unnonouno** [comment]
This issue is related to #484

**Comment by beam2d** [comment]
Oh, thank you for the catch!

**Comment by tscohen** [comment]
I agree this would be very useful. Perhaps an efficient GPU backward kernel can be written that works something like this: 
- Launch a thread for each cell in the *output* of the indexing operation
- Determine the index i in the input array that was used in the forward pass to get the value for the output cell processed by this thread. This may involve fancy indexing computations, but is the same code as used in fprop.
- Use CUDA's atomicAdd() to add the output_grad of this thread's cell to the input_grad cell i.

**Comment by entron** [comment]
As this has not been implemented yet, what can I do if I want slicing? Is `get_item` function equivalent way (though much more verbose) to do slicing?

**Comment by wkentaro** [comment]
@entron You can do like this. https://github.com/pfnet/chainer/blob/master/tests/chainer_tests/test_variable.py#L66-L76

**Comment by entron** [comment]
@wkentaro Thanks for the suggestion! That simplified the syntax a bit to:

```
x[slice(None), slice(1, None)]
```

Still it is not as elegant as the simple numpy syntax. Hoping this issue will be implement soon!

**Comment by wkentaro** [comment]
@entron It also works.
```
>>> import numpy as np
>>> import chainer
>>> x_data = np.arange(35).reshape((5, 7))
>>> x = chainer.Variable(x_data)
>>> x.data.shape
(5, 7)
>>> x[:, 1:].data.shape
(5, 6)
```

**Comment by wkentaro** [comment]
For the numpy-like advanced indexing, I have tried to work for this issue by implementing boolean_array_indexing kernel function.
But I encountered a problem while implementing the kernel function as described in [the forum](https://groups.google.com/forum/#!topic/chainer/O1kswVMCY98), and I have not solved the issue yet.

**Comment by entron** [comment]
I see, indeed the numpy style indexing works. It is just not backprop-able. If one want to build a function then one has to use `get_item ` right? One thing I find probably dangerous is that there is no warning indicating that numpy style indexing is not backprop-able. This could be very difficult to debug if one does not aware of this issue.

**Comment by wkentaro** [comment]
>  If one want to build a function then one has to use get_item right? 

No, you can use that indexing. `x[:, 1:]` as I do in [Fully Convolutional Networks](https://github.com/wkentaro/fcn/blob/master/fcn/models/fcn8s.py#L132-L134).
With these line:
- https://github.com/pfnet/chainer/blob/master/chainer/__init__.py#L80
- https://github.com/pfnet/chainer/blob/master/chainer/functions/array/get_item.py#L49-L50

**Comment by entron** [comment]
Do you mean I have to copy those lines and pasted in the beginning of my code? Sorry for these might be obvious questions:)

**Comment by wkentaro** [comment]
Maybe there are some misunderstandings, but I just mean

- numpy-like basic indexing for chainer.Variable does work and backprop-able: `x[:, 1:]`
- `get_item` function does work and backprop-able: `F.get_item(x, (slice(None), slice(1, None)))`
- numpy-like advanced indexing for chainer.Variable does not work and not backprop-able: `x[x.data == 0]`

so maybe what you're pointing below does not occur.

> I see, indeed the numpy style indexing works. It is just not backprop-able. If one want to build a function then one has to use get_item right? One thing I find probably dangerous is that there is no warning indicating that numpy style indexing is not backprop-able, which could be very difficult to debug if one does not aware of this issue.

**Comment by entron** [comment]
Now I understand! Thank you very much!

**Comment by tscohen** [comment]
Looks like this can be closed (implemented in #1369)

**Comment by beam2d** [comment]
As commented above, this issue is resolved.
